src/routers/search_router.py

```
@@ -0,0 +1,181 @@
+from pydantic import BaseModel
+from sqlalchemy.engine import Engine
+from starlette import status
+
+from authentication import get_current_user#, has_role
+from database.model.concept.aiod_entry import AIoDEntryRead
+from database.model.resource_read_and_create import resource_read
+
+SORT = {"identifier": "asc"}
+LIMIT_MAX = 1000
+
+RESOURCE = TypeVar("RESOURCE")
+
+
+class SearchResult(BaseModel, Generic[RESOURCE]):
+    total_hits: int
+    resources: list[RESOURCE]
+    next_offset: list | None
+    current_page: int
+    page_size: int
+
+class SearchRouter(Generic[RESOURCE], abc.ABC):
+    """
+    Providing search functionality in ElasticSearch
+    """
+    
+    def __init__(self, client: Elasticsearch):
+        self.client: Elasticsearch = client
+    
+    @property
+    @abc.abstractmethod
+    def es_index(self) -> str:
+        """The name of the elasticsearch index"""
+    
+    @property
+    @abc.abstractmethod
+    def resource_name_plural(self) -> str:
+        """The name of the resource (plural)"""
+    
+    @property
+    def key_translations(self) -> dict[str, str]:
+        """If an attribute is called differently in elasticsearch than in our
+        metadata model, you can define a translation dictionary here. The key
+        should be the name in elasticsearch, the value the name in our data
+        model."""
+        return {}
+    
+    @property
+    @abc.abstractmethod
+    def resource_class(self) -> RESOURCE:
+        """The resource class"""
+    
+    def create(self, engine: Engine, url_prefix: str) -> APIRouter:
+        router = APIRouter()
+        read_class = resource_read(self.resource_class)  # type: ignore
+        
+        # TODO: check parameters correctness
+        @router.get(f"{url_prefix}/search/{self.resource_name_plural}/v1",
+                    tags=["search"])
+        def search(
+            platforms: Annotated[list[str] | None, Query()] = None,
```

Do you want to raise an exception on non-existent platforms?

---

For that kind of decision we really need to have some talk with everyone that needs the searches, because I actually only saw the marketplace, which has the platforms as a selection list... but maybe from other places they will allow to write them, I don't know

---

In any case, we could raise the exception, but it also imply something... Will we retrieve them from the database to compare with? If we don't, every time we add a new one, we will need to actualise this code too

---

The platforms are already stored in the db